### PR TITLE
Adds support for energy based learning with NLL loss (LEO)

### PR DESCRIPTION
**Pull request by psodhi**
This PR introduces LEO, learning energy-based models in optimization (https://arxiv.org/abs/2108.02274). LEO is a method to learn models end-to-end within second-order optimizers like Gauss-Newton. The main difference is that instead of unrolling the optimizer and minimizing the MSE tracking loss, this introduces the NLL energy-based loss that does not backpropagate through the optimizer. It requires low-energy samples from the optimizer, pushing up the energy of optimizer samples and pushing down the energy of ground truth samples.

To execute it, run
`python examples/state_estimation_2d.py` with `learning_method="leo"`

This should update the learnable cost weights so that the optimizer trajectory (orange) matches the ground truth trajectory (green).

### CODE REVIEW

File: theseus/optimizer/nonlinear/gauss_newton.py [hunk 1]
```diff
@@ -41,3 +41,29 @@ def __init__(
 
     def compute_delta(self, **kwargs) -> torch.Tensor:
         return self.linear_solver.solve()
+
+    def compute_samples(self, n_samples: int = 10, T: float = 1.0) -> torch.Tensor:
+        delta = self.linear_solver.solve()
+        AtA = self.linear_solver.precision() / T
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/gauss_newton.py:47]: Any reason to have a separate `precision` function call since it just return `AtA` which you could directly access here?

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py [hunk 2]
```diff
@@ -220,6 +220,9 @@ def _optimize_impl(
     def compute_delta(self, **kwargs) -> torch.Tensor:
         pass
 
+    def compute_samples(self, n_samples: int, T: float) -> torch.Tensor:
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/nonlinear_optimizer.py:223]: Leave a comment here as discussed to say what's the intended use, in case it's confusing for a user to see a sampling function inside an optimizer.

File: theseus/optimizer/nonlinear/gauss_newton.py [hunk 3]
```diff
@@ -41,3 +41,29 @@ def __init__(
 
     def compute_delta(self, **kwargs) -> torch.Tensor:
         return self.linear_solver.solve()
+
+    def compute_samples(self, n_samples: int = 10, T: float = 1.0) -> torch.Tensor:
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/gauss_newton.py:45]: I am guessing `T` is temperature? Maybe pick a more descriptive name or describe in comments. Same for `B`, `N`, `R` below, where you can point to the equations from any reference.

File: theseus/optimizer/nonlinear/levenberg_marquardt.py [hunk 4]
```diff
@@ -65,3 +65,29 @@ def compute_delta(
             ellipsoidal_damping=ellipsoidal_damping,
             damping_eps=damping_eps,
         )
+
+    def compute_samples(self, n_samples: int = 10, T: float = 1.0) -> torch.Tensor:
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/levenberg_marquardt.py:69]: This is identical to the one in gauss newton. Maybe pull this up to nonlinear least squares class instead?

File: theseus/optimizer/nonlinear/gauss_newton.py [hunk 5]
```diff
@@ -41,3 +41,29 @@ def __init__(
...
+        delta = self.linear_solver.solve()
+        AtA = self.linear_solver.precision() / T
+        R = torch.linalg.cholesky(AtA, upper=True)
+
+        B, N = delta.shape
+        y = torch.normal(
+            mean=torch.zeros((N, n_samples), device=delta.device),
+            std=torch.ones((N, n_samples), device=delta.device),
+        )
+        delta_samples = (torch.triangular_solve(y, R).solution) + (
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/gauss_newton.py:55]: Add a comment for reference to the math.

File: theseus/tests/test_theseus_layer.py [hunk 6]
```diff
@@ -280,11 +301,44 @@ def cost_weight_fn():
...
+        if learning_method == "leo":
+            # groundtruth cost
+            x_gt = target_vars["coefficients"]
+            input_values_gt = {
+                "coefficients": x_gt,
+                cost_weight_param_name: cost_weight_fn(),
+            }
+            layer_to_learn.objective.update(input_values_gt)
+            cost_gt = torch.sum(layer_to_learn.objective.error(), dim=1)  # B x 1
+            cost_gt = cost_gt.to(device)
```
Comment (mhmukadam) [inline theseus/tests/test_theseus_layer.py:313]: Might not be the best place to move between device in the learning loop. This line won't be needed if inputs, objective, layer etc are setup to be on the correct device at the beginning.

File: theseus/tests/test_theseus_layer.py [hunk 7]
```diff
@@ -280,11 +301,44 @@ def cost_weight_fn():
...
+            loss = torch.mean(loss, dim=0)
+        else:
+            loss = F.mse_loss(pred_vars["coefficients"], target_vars["coefficients"])
+
         loss.backward()
-        print(i, loss.item(), loss.item() / loss0)
         optimizer.step()
-        if loss.item() / loss0 < 5e-3:
+
+        mse_loss = F.mse_loss(pred_vars["coefficients"], target_vars["coefficients"])
```
Comment (mhmukadam) [inline theseus/tests/test_theseus_layer.py:339]: Instead of calculating the mse loss again you can shuffle things around.

```
mse_loss = F.mse_loss(pred_vars["coefficients"], target_vars["coefficients"])
....
if learning_method == "leo":
...
else:
    loss = mse_loss

loss.backward()
optimizer.step()

print(i, mse_loss.item())        
if mse_loss.item() / loss0 < 5e-3:
    solved = True
    break
...
```

File: examples/state_estimation_2d.py [hunk 8]
```diff
@@ -236,21 +306,60 @@ def cost_weights_model():
                 print_stuff=epoch % 10 == 0 and i == 0,
             )
 
-        solution_path = get_path_from_values(
-            objective.batch_size, theseus_inputs, path_length
-        )
+        if learning_method == "leo":
```
Comment (mhmukadam) [inline examples/state_estimation_2d.py:309]: Add comment as discussed about high level explanation of LEO and reference to the paper and any specific equations.

**Review by mhmukadam — COMMENTED** [review]
Looks great! The API looks good -- just a sampling function in the optimizer class. Added some comments to address. The GPU test is failing due to cholesky upper not available.  We can add support for sparse solvers and optimize performance in later PRs.

**Review by luisenp — CHANGES_REQUESTED** [review]
Great job on this PR! I see all the functionality is there, but I propose we shuffle some things around a bit. I'm OK leaving that as a separate PR, but we should definitely make issues for the comments so we don't forget. 

Regarding the broken test, `upper=` kwarg for cholesky was only introduced in `torch` 1.10.0. It used to be in `torch.cholesky`, but it was removed in  `torch.linalg.cholesky` in v1.9.0 (see screenshot). 

For now, can you update the call to match the one used in v1.9.0 so that the test passes?
image

File: theseus/optimizer/nonlinear/levenberg_marquardt.py (hunk 4 above)
Comment (psodhi) [reply theseus/optimizer/nonlinear/levenberg_marquardt.py:69]: Good point, I'll move this up to nonlinear least squares for now. The main consideration is if the same function is applicable for future least-squares solvers that would derive from this class.

Comment (mhmukadam) [reply theseus/optimizer/nonlinear/levenberg_marquardt.py:69]: This proposal is much cleaner so if the logic works generally (say for sparse linear solvers and other possible future solvers) then let's go with this.

Do remember you will need to have run `TheseusLayer.forward` first before calling `TheseusLayer.compute_samples`.

File: examples/state_estimation_2d.py [hunk 9]
```diff
@@ -120,6 +154,42 @@ def get_path_from_values(batch_size_, values_, path_length_):
...
+    :return: values: dict of (x,y) pos values
+    """
+    [batch_size_, path_length_, dim] = path_.shape
+    values = {}
+    for i in range(path_length_):
+        values[f"pose_{i}"] = path_[:, i, :2]
+    return values
+
+
+def get_average_sample_cost(x_samples, cost_weights_model, objective, mode_):
```
[the remark this answers is not in the document]
Comment (psodhi) [reply examples/state_estimation_2d.py:169]: Agreed, sounds good.

File: examples/state_estimation_2d.py (hunk 8 above)
Comment (psodhi) [reply examples/state_estimation_2d.py:313]: Added comments.

File: theseus/tests/test_theseus_layer.py (hunk 6 above)
Comment (psodhi) [reply theseus/tests/test_theseus_layer.py:313]: Agreed, removed the to(device)

File: theseus/tests/test_theseus_layer.py [hunk 10]
```diff
@@ -280,11 +301,44 @@ def cost_weight_fn():
...
+            layer_to_learn.objective.update(input_values_gt)
+            cost_gt = torch.sum(layer_to_learn.objective.error(), dim=1)  # B x 1
+            cost_gt = cost_gt.to(device)
+
+            # optimizer cost
+            x_opt = pred_vars["coefficients"].detach()
+            x_samples = layer_to_learn.optimizer.compute_samples(
+                n_samples=10, T=1.0
+            )  # B x N x S
+            if x_samples is None:  # use mean solution
```
[the remark this answers is not in the document]
Comment (psodhi) [reply theseus/tests/test_theseus_layer.py:320]: It is possible when learning_method == "leo" but the underlying optimizer doesn't define a sampling function. In this case, it reverts to an MMP algorithm where it uses the mean trajectory.

File: theseus/tests/test_theseus_layer.py (hunk 7 above)
Comment (psodhi) [reply theseus/tests/test_theseus_layer.py:339]: Agreed, moved for both test and state_estimation example.

File: theseus/optimizer/nonlinear/levenberg_marquardt.py (hunk 4 above)
Comment (psodhi) [reply theseus/optimizer/nonlinear/levenberg_marquardt.py]: On thinking about it more, it does make sense for `compute_samples()` to be independent of the optimizer. I have moved `compute_samples()` to theseus_layer where it uses the current linearization point. However, instead of `linearization`, it is using `linear_solver` as input because we need to compute delta as well within the function.

File: theseus/optimizer/nonlinear/gauss_newton.py (hunk 1 above)
Comment (psodhi) [reply theseus/optimizer/nonlinear/gauss_newton.py]: Moved it to `Linearization` where it is called `hessian_approx()`.

File: theseus/optimizer/nonlinear/levenberg_marquardt.py (hunk 4 above)
Comment (psodhi) [reply theseus/optimizer/nonlinear/levenberg_marquardt.py]: Yes the final computation is a similar functionality as `retract_and_update_variables`.  One important issue to be flagged for the future is to broadcast the `retract_and_update_variables`. It would make sampling much faster.

File: theseus/tests/test_theseus_layer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mhmukadam) [reply theseus/tests/test_theseus_layer.py:321]: Should we add an or to cover x_samples <= 0?

File: theseus/theseus_layer.py [hunk 11]
```diff
@@ -45,6 +46,39 @@ def forward(
...
         return values, info
 
+    def compute_samples(
+        self,
+        linear_solver: LinearSolver = None,
+        n_samples: int = 10,
+        temperature: float = 1.0,
+    ) -> torch.Tensor:
+        if linear_solver is None:
+            return None
```
Comment (mhmukadam) [inline theseus/theseus_layer.py:56]: raise error?

File: theseus/theseus_layer.py [hunk 12]
```diff
@@ -45,6 +46,39 @@ def forward(
...
+        delta = linear_solver.solve()
```
Comment (mhmukadam) [inline theseus/theseus_layer.py:57]: Is there a way to avoid solving the linear system again? This should be computed in the last iteration in forward.

File: theseus/optimizer/nonlinear/levenberg_marquardt.py (hunk 4 above)
Comment (mhmukadam) [reply theseus/optimizer/nonlinear/levenberg_marquardt.py]: Sounds good, I'll flag an issue for this.

**Review by mhmukadam — APPROVED** [review]

File: theseus/optimizer/nonlinear/gauss_newton.py (hunk 5 above)
Comment (psodhi) [reply theseus/optimizer/nonlinear/gauss_newton.py]: Added a reference.

File: theseus/theseus_layer.py (hunk 11 above)
Comment (psodhi) [reply theseus/theseus_layer.py:56]: Returned None here since if learning_method is set to "leo" and there isn't a valid sampling function, this would revert to a default MMP algorithm where it uses the mean trajectory. I can change it to a RaiseError if that seems better.

Comment (mhmukadam) [reply theseus/theseus_layer.py:56]: Actually this is better. We should just document this in the comments.

Comment (psodhi) [reply theseus/theseus_layer.py:58]: Sounds good, documented in the comments.

**Review by mhmukadam — APPROVED** [review]
